col/col.1
.\"
.\" Derived from col(1), Unix 32V:
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\"
.\" SPDX-Licence-Identifier: Caldera
.\"
.TH COL 1 "5/26/03" "Heirloom Toolchest" "User Commands"
.SH NAME
col \- filter reverse line feeds
.SH SYNOPSIS
\fBcol\fR [\fB\-bfxp\fR]
.SH DESCRIPTION
.I Col
reads the standard input and writes the standard output.
It performs the line overlays implied by reverse line
feeds (ESC-7 in ASCII)
and by forward and reverse half line feeds (ESC-9 and ESC-8).
.I Col
is particularly useful for filtering multicolumn
output made with the `.rt' command of
.I nroff
and output resulting from use of the
.IR tbl (1)
preprocessor.
.PP
Although
.I col
accepts half line motions in its input, it normally does not
emit them on output.
Instead, text that would appear between lines is moved to the next lower
full line boundary.
This treatment can be suppressed by the
.B \-f
(fine) option; in this case
the output from
.I col
may contain forward half line feeds (ESC-9), but will still never contain
either kind of reverse line motion.
.PP
If the
.B \-b
option is given,
.I col
assumes that the output device in use is not capable of backspacing.
In this case, if several characters are to appear in the same place,
only the last one read will be taken.
.PP
The control characters SO (ASCII code 017),
and SI (016) are assumed
to start and end text in an alternate character set.
The character set (primary or alternate) associated with each printing
character read is remembered; on output, SO and SI characters are generated
where necessary to maintain the correct treatment of each character.
.PP
.I Col
normally converts white space to tabs to shorten printing time.
If the
.B \-x
option is given, this conversion is suppressed.
.PP
All control characters are removed from the input except space,
backspace,
tab, return, newline, ESC (033) followed by one of 7, 8, 9, SI, SO, and VT
(013).
This last character is an alternate form of full reverse line feed, for
compatibility with some other hardware conventions.
With the
.B \-p
option,
all escape sequences are passed through,
but
.I col
does not perform any interpretation of sequences unknown to it
so its output may not be appropriate with this option.
All other non-printing characters are ignored.
.SH "ENVIRONMENT VARIABLES"
.TP
.BR LANG ", " LC_ALL
See
.IR locale (7).
.TP
.B LC_CTYPE
Determines the mapping of bytes to characters,
the set of printable characters,
and the width of characters.
.SH "SEE ALSO"
troff(1), tbl(1)
.SH NOTES
Can't back up more than 256 lines.


CHANGES
Release ...
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants. 

Release 230503 
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin. 

Release 230120
* Notes for the library directories if linking against the GNU C library.
* rm: Consistency within standards: the SUS program shouldn't have the POSIX 1003.1(2008) fix, just the main program.
* Made $(ROOT) not hardcoded anymore.
* Cleaned up more licence headers (if my estimations are correct, we've got rid of 3.255 lines!).
* Small update on README.md --- although it's now less readable on plain-text readers such as more or pg.

Release 230107
* rm: Added support for the POSIX 1003.1(2008).
* chroot: Added manual page.
* awk: Added stdint.h to awk.g.y.
* Now it has a bigger range of compatibility with GNU C library versions.

Release 221025
* Made notes clearer about building Heirloom on modern versions of the GNU C library.
* tar: Support for zstd and xz.

Release 220821
* [NEW] chroot: New implementation of chroot, based on Solaris 2.5 payloads.
* true, false: Rewritten in C.
* yes: Rewritten and adapted to ANSI C.
* "libcommon/vpfmt.c": Adapted to ANSI C.
* Now many of the utilities can be built in the musl C library and GNU C library (dynamically).

Release 070715
* cpio: A "-P" option has been introduced that allows to specify attributes
  like the file owner when creating an archive. With it, unprivileged users
  can set arbitrary attributes for files within an archive.
* diff: Made it compile on CentOS 5.
* Static linking against glibc is now possible since "optopt" is no longer
  a common symbol, and there is no conflict anymore with "getopt.o".
* cpio, tar: A compilation problem on some Linux systems has been fixed.
* ul: A compilation problem for NetBSD and other systems that use termcap
  instead of curses has been fixed (patch by Aleksey Cheusov).
* bfs: An OpenSolaris-derived manual page is now included.

Release 070227
* The directory for development utilities, by default "/usr/ccs/bin", is
  now configurable in "build/mk.config" . It is expected that the utilities
  from the "Heirloom Development Tools" package are installed in it.
* getconf: "/usr/ccs/bin" is now included in the default path as printed
  with the "PATH" argument.
* what: Has been replaced by an OpenSolaris-derived implementation.
* tsort: Is now installed in "/usr/ccs/bin".
* OpenSolaris-derived manual pages are now included for bdiff, newform,
  and fspec(5).
* psrinfo: Fixed the computation of the "on-line since" time for Linux 2.6.

col/col.c
/*
 * Changes by Gunnar Ritter, Freiburg i. Br., Germany, May 2003.
 */
/*	from Unix 32V /usr/src/cmd/col.c	*/
/*
 * Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
 *
 * SPDX-Licence-Identifier: Caldera
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
static const char sccsid[] USED = "@(#)col.sl	1.10 (gritter) 6/28/10";

# include <stdio.h>
# include <stdlib.h>
# include <libgen.h>
# include <string.h>
# include <locale.h>
# include <ctype.h>
# include <wchar.h>
# include <wctype.h>
# include <unistd.h>
# include <iblok.h>
# if defined (__GLIBC__) && defined (_IO_putc_unlocked)
# undef putchar
# define putchar(c)	_IO_putc_unlocked(c, stdout)
# endif
# define PL 256
# define ESC '\033'
# define RLF '\013'
# define SI '\017'
# define SO '\016'
# define CMASK	0x00000000ffffffffLL
# define EOFC	0x0000010000000000LL
# define TRANS	0x1000000000000000LL
# define GREEK	0x2000000000000000LL
# define ILLSEQ	0x4000000000000000LL
# define ISPRNT	0x8000000000000000LL
# define COLMSK	0x0f00000000000000LL
# define COLSHF	56
# define CHRMSK 0x000000ff00000000LL
# define CHRSHF	32

# define nextc(c, cp, m)	(mb_cur_max > 1 ? \
		((cp) = ib_getw(input, &(c), &(m)), \
		 (cp) == NULL ? EOFC : \
		 (c) == WEOF ? ((long long)(*(cp)&0377)<<CHRSHF)|ILLSEQ : \
		 iswprint(c) ? (c)|((long long)wcwidth(c)<<COLSHF)|ISPRNT : \
		 (c)) : \
		((c) = ib_get(input), \
		 (c) == EOF ? EOFC : \
		 isprint(c) ? (c)|(1LL<<COLSHF)|ISPRNT : \
		 (c)))

# define slbuff(n)	(line+(n)+128 >= LINELN ? growsbuff((n)+128) : lbuff);

static long long *page[PL];
static long long *lbuff;
static long	line, LINELN;
static int bflag, xflag, fflag, pflag;
static int half;
static int cp, lp;
static int ll, llh, mustwr;
static int pcp = 0;
static char *pgmname;
static struct iblok	*input;
static int	mb_cur_max;

static long long	space[] = { ' ' | (1LL<<COLSHF) | ISPRNT, '\0' };

static void	outc(long long);
static void	store(int);
static void	fetch(int);
static void	emit(long long *, int);
static void	incr(void);
static void	decr(void);
static size_t	lllen(const long long *);
static long long	*llcpy(long long *, const long long *);
static long long	*growsbuff(size_t);

int
main (int argc, char **argv)
{
	int i;
	long long greek;
	long long c;
	wint_t	wc;
	char	*mp;

	pgmname = basename(argv[0]);
	setlocale(LC_CTYPE, "");
	mb_cur_max = MB_CUR_MAX;

	while ((i = getopt(argc, argv, ":bfxp")) != EOF) {
		switch (i) {
		case 'b':
			bflag++;
			break;
		case 'f':
			fflag++;
			break;
		case 'x':
			xflag++;
			break;
		case 'p':
			pflag++;
			break;
		default:
			fprintf (stderr, "%s: bad option letter %c\n",
					pgmname, optopt);
				exit (2);
		}
	}
	if (optind < argc) {
		fprintf (stderr, "%s: bad option %s\n",
				pgmname, argv[optind]);
		exit (2);
	}

	if ((input = ib_alloc(0, 0)) == NULL) {
		perror(pgmname);
		exit(1);
	}
	for (ll=0; ll<PL; ll++)
		page[ll] = 0;

	cp = 0;
	ll = 0;
	greek = 0;
	mustwr = PL;
	line = 0;
	slbuff(0);

	while ((c = nextc(wc, mp, i)) != EOFC) {
		switch (c & CMASK) {
		case '\n':
			incr();
			incr();
			cp = 0;
			continue;

		case '\0':
			continue;

		case ESC:
			c = nextc(wc, mp, i);
			switch (c & CMASK) {
			case '7':	/* reverse full line feed */
				decr();
				decr();
				break;

			case '8':	/* reverse half line feed */
				if (fflag)
					decr();
				else {
					if (--half < -1) {
						decr();
						decr();
						half += 2;
					}
				}
				break;

			case '9':	/* forward half line feed */
				if (fflag)
					incr();
				else {
					if (++half > 0) {
						incr();
						incr();
						half -= 2;
					}
				}
				break;
			default:
				if (pflag) {
					outc(ESC);
					cp++;
					if (c != EOFC) {
						outc(c);
						cp += (c&COLMSK)>>COLSHF;
					}
				}
			}
			continue;

		case SO:
			greek = GREEK;
			continue;

		case SI:
			greek = 0;
			continue;

		case RLF:
			decr();
			decr();
			continue;

		case '\r':
			cp = 0;
			continue;

		case '\t':
			cp = (cp + 8) & -8;
			continue;

		case '\b':
			if (cp > 0)
				cp--;
			continue;

		case ' ':
			cp++;
			continue;

		default:
			if (c & ISPRNT) {/* if printable */
				outc(c | greek);
				cp += (c&COLMSK)>>COLSHF;
			}
			continue;
		}
	}

	for (i=0; i<PL; i++)
		if (page[(mustwr+i)%PL] != 0)
			emit (page[(mustwr+i) % PL], mustwr+i-PL);
	emit (space, (llh + 1) & -2);
	return 0;
}

static void
outc (register long long c)
{
	int i, v, w;
	size_t n;

	if (lp > cp) {
		line = 0;
		lp = 0;
	}

	while (lp < cp) {
		slbuff(0);
		switch (lbuff[line]) {
		case '\0':
			lbuff[line] = ' ' | ISPRNT | (1LL<<COLSHF);
			lp++;
			break;

		case '\b':
			lp--;
			break;

		default:
			lp += (lbuff[line]&COLMSK)>>COLSHF;
			if (lp > cp)
				continue;
		}
		line++;
	}
	slbuff(0);
	while ((lbuff[line] & CMASK) == '\b') {
		line += 2;
		slbuff(0);
	}
	if (bflag || lbuff[line] == '\0' || (lbuff[line]&CMASK) == ' ' &&
			(lbuff[line]&TRANS) == 0) {
		w = (lbuff[line]&COLMSK)>>COLSHF;
		if (lp > cp) {
			w -= lp-cp;
			n = lllen(&lbuff[line]) + 1;
			slbuff(lp-cp+n);
			memmove(&lbuff[line+lp-cp], &lbuff[line],
					n * sizeof *lbuff);
			n = lp - cp;
			while (n--)
				lbuff[line++] = ' '|ISPRNT|(1LL<<COLSHF);
			lp = cp;
		}
		v = (c&COLMSK)>>COLSHF;
		lbuff[line] = c;
		if (w > v) {
			n = lllen(&lbuff[line+w-v]) + 1;
			slbuff(w-v+n);
			memmove(&lbuff[line+w-v], &lbuff[line],
					n * sizeof *lbuff);
			n = w - v;
			while (n--)
				lbuff[++line] = ' '|ISPRNT|(1LL<<COLSHF);
		} else {
			n = 1;
			while (lbuff[line+n] = lbuff[line+v-w+n])
				n++;
		}
	} else {
		if (lp > cp)
			line++;
		slbuff(1);
		w = (lbuff[line++]&COLMSK)>>COLSHF;
		if (lp > cp) {
			w += lp - cp;
			lp = cp;
		}
		v = (c&COLMSK)>>COLSHF;
		n = lllen(&lbuff[line]) + 1;
		slbuff(w+1+n);
		memmove(&lbuff[line+w+1], &lbuff[line],
				n * sizeof *lbuff);
		for (i = 0; i < w; i++) {
			slbuff(0);
			lbuff[line++] = '\b';
		}
		slbuff(0);
		lbuff[line++] = c;
		if (w > v) {
			n = lllen(&lbuff[line+w-v]) + 1;
			slbuff(w-v+n);
			memmove(&lbuff[line+w-v], &lbuff[line],
					n * sizeof *lbuff);
			n = w - v;
			while (n--)
				lbuff[++line] = ' '|ISPRNT|(1LL<<COLSHF)|TRANS;
		} else {
			n = 1;
			while (lbuff[line+n] = lbuff[line+v-w+n])
				n++;
		}
		lp = 0;
		line = 0;
	}
}

static void
store (int lno)
{
	lno %= PL;
	if (page[lno] != 0)
		free (page[lno]);
	page[lno] = malloc((lllen(lbuff) + 2) * sizeof *page[lno]);
	if (page[lno] == 0) {
		fprintf (stderr, "%s: no storage\n", pgmname);
		exit (2);
	}
	llcpy (page[lno],lbuff);
}

static void
fetch(int lno)
{
	register long long *p;
	size_t	n;

	lno %= PL;
	p = lbuff;
	while (*p)
		*p++ = '\0';
	line = 0;
	lp = 0;
	if (page[lno]) {
		n = lllen(page[lno]);
		slbuff(n + 1);
		llcpy (&lbuff[line], page[lno]);
	}
}
static void
emit (long long *s, int lineno)
{
	static int cline = 0;
	register int ncp, i;
	register long long *p;
	static long long gflag = 0;

	if (*s) {
		while (cline < lineno - 1) {
			putchar ('\n');
			pcp = 0;
			cline += 2;
		}
		if (cline != lineno) {
			putchar (ESC);
			putchar ('9');
			cline++;
		}
		if (pcp)
			putchar ('\r');
		pcp = 0;
		p = s;
		while (*p) {
			ncp = pcp;
			while ((*p++&CMASK) == ' ') {
				if ((++ncp & 7) == 0 && !xflag) {
					pcp = ncp;
					putchar ('\t');
				}
			}
			if (!*--p)
				break;
			while (pcp < ncp) {
				putchar (' ');
				pcp++;
			}
			if (gflag != (*p & GREEK) && (*p&CMASK) != '\b') {
				if (gflag)
					putchar (SI);
				else
					putchar (SO);
				gflag ^= GREEK;
			}
			if (*p & ILLSEQ)
				putchar((*p&CHRMSK)>>CHRSHF);
			else if (mb_cur_max == 1)
				putchar(*p & 0377);
			else {
				char	mb[MB_LEN_MAX];
				int	n;
				n = wctomb(mb, *p&CMASK);
				for (i = 0; i < n; i++)
					putchar(mb[i] & 0377);
			}
			if ((*p&CMASK) == '\b') {
				pcp--;
			} else
				pcp += (*p&COLMSK)>>COLSHF;
			p++;
		}
	}
}

static void
incr(void)
{
	store (ll++);
	if (ll > llh)
		llh = ll;
	if (ll >= mustwr && page[ll%PL]) {
		emit (page[ll%PL], ll - PL);
		mustwr++;
		free (page[ll%PL]);
		page[ll%PL] = 0;
	}
	fetch (ll);
}

static void
decr(void)
{
	if (ll > mustwr - PL) {
		store (ll--);
		fetch (ll);
	}
}

static size_t
lllen(const long long *lp)
{
	size_t	n = 0;

	while (*lp++)
		n++;
	return n;
}

static long long
*llcpy(long long *dst, const long long *sp)
{
	long long	*dp = dst;

	while (*dp++ = *sp++);
	return dst;
}

static long long *
growsbuff(size_t n)
{
	if ((lbuff = realloc(lbuff, (line+n) * sizeof *lbuff)) == NULL) {
		write(2, "no memory\n", 10);
		_exit(077);
	}
	memset(&lbuff[LINELN], 0, (line+n-LINELN)*sizeof *lbuff);
	LINELN = line + n;
	return lbuff;
}
